cellengine/resources/compensation.py

```diff
@@ -100,30 +101,38 @@ def dataframe_as_html(self):
         """Return the compensation matrix dataframe as HTML."""
         return self.dataframe._repr_html_()
 
-    def apply(self, file: "FcsFile", inplace: bool = False, **kwargs):
-        """
-        Compensate an FcsFile's data.
+    def apply(self, file: FcsFile, inplace: bool = True, **kwargs):
+        """Compensate an FcsFile's data.
 
         Args:
-            file (FcsFile): The FCS file to compensate.
-            inplace (bool): Compensate the file's data in-place.
+            file (FcsFile): The FcsFile to compensate.
+            inplace (bool): If True, modify the `FcsFile.events` with the result.
+                If False, return the compensated events.
             kwargs (Dict):
                 All arguments accepted by `FcsFile.get_events` are accepted here.
         Returns:
             DataFrame: if ``inplace=True``, updates `FcsFile.events` for
                 the target FcsFile
         """
-        data: DataFrame = file.get_events(
-            **kwargs, inplace=True, destination=None
-        )  # type: ignore
+        data = file.get_events(inplace=inplace, destination=None, **kwargs)
 
         # spill -> comp by inverting
-        inverted = numpy.linalg.inv(self.dataframe)
+        if self._inverted:
+            inverted = self._inverted
+        else:
+            inverted = numpy.linalg.inv(self.dataframe)
+            self._inverted = inverted
```

Comment:
Gahh one last thing. When `self.dataframe` changes, `self._inverted` needs to be invalidated.